### PR TITLE
[pickers] Adds RTL verification for the time pickers default views 

### PR DESCRIPTION
**Pull request by arthurbalduini**
Closes [#11525](https://github.com/mui/mui-x/issues/11525)

### CODE REVIEW

File: packages/x-date-pickers/src/TimePicker/shared.tsx [hunk 1]
```diff
@@ -109,7 +112,7 @@ export function useTimePickerDefaultizedProps<
     ...applyDefaultViewProps({
       views: themeProps.views,
       openTo: themeProps.openTo,
-      defaultViews: ['hours', 'minutes'] as TView[],
+      defaultViews: defaultViews as TView[],
```
Comment (LukasTy) [inline packages/x-date-pickers/src/TimePicker/shared.tsx:115]: This would only fix cases, where default `views` values are used.
It would also not impact the `DateTimePicker` or any other place the `MultiSectionDigitalClock` is used. 🤔 
Screenshot 2024-06-12 at 10 25 32

**Review by LukasTy — COMMENTED** [review]
It should also apply to the plain `MultiSectionDigitalClock`.
Screenshot 2024-06-12 at 10 35 02

**Comment by LukasTy** [comment]
I feel like it should be doable within `MultiSectionDigitalClock` using style adjustments.
Have you considered a `flex order`?

File: packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx [hunk 2]
```diff
@@ -143,10 +145,16 @@ export const MultiSectionDigitalClock = React.forwardRef(function MultiSectionDi
...
     }
-    return inViews.includes('meridiem') ? inViews : [...inViews, 'meridiem'];
-  }, [ampm, inViews]);
+
+    if (!isRtl) {
+      return inViews.includes('meridiem') ? inViews : [...inViews, 'meridiem'];
+    }
+
+    const viewsWithoutMeridiem = inViews.filter((view) => view !== 'meridiem');
+    return viewsWithoutMeridiem.toReversed().concat('meridiem');
```
Comment (LukasTy) [inline packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx:156]: We are playing with fire here regarding views. 🙈 
Making such structural `views` adjustments is tricky and this is the place where such change creates the most problems. 🙈 
The tabulation/views process does not work as expected in RTL mode.
Given a demo with default settings, the minutes view is skipped—selecting hours jumps the focus to meridiem and after selecting it the Picker closes.

Have you tried adjusting only the views intended for rendering?

**Comment by Eliav2** [comment]
Dude I'm not sure what you changed there but it looks overkill
Want me to make a PR ?

**Comment by LukasTy** [comment]
> Dude I'm not sure what you changed there but it looks overkill
> Want me to make a PR ?

If you have ideas on how to do it—the suggestion are welcome.
I imagine, that given your suggestion, all the time sections, except for meridiem should be wrapped in a single element and then your suggestion should apply.
Screenshot 2024-06-17 at 20 21 01

Or do you @Eliav2 have a suggestion on how to achieve it without a breaking DOM change? 🤔

File: packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx [hunk 3]
```diff
@@ -390,13 +392,24 @@ export const MultiSectionDigitalClock = React.forwardRef(function MultiSectionDi
   );
 
   const viewTimeOptions = React.useMemo(() => {
-    return views.reduce(
-      (result, currentView) => {
-        return { ...result, [currentView]: buildViewProps(currentView) };
+    // Filter out the 'meridiem' view and orders the other views
+    const otherViews = views.filter((v) => v !== 'meridiem');
```
Comment (LukasTy) [inline packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx:396]: I noticed that we rely on the object having the expected order of items, but that is risky in JS.
WDYT instead about the following diff:

```diff
diff --git a/packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx b/packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx
index ace4ba50b..8e396919a 100644
--- a/packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx
+++ b/packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx
@@ -391,25 +391,26 @@ export const MultiSectionDigitalClock = React.forwardRef(function MultiSectionDi
     ],
   );
 
-  const viewTimeOptions = React.useMemo(() => {
-    // Filter out the 'meridiem' view and orders the other views
-    const otherViews = views.filter((v) => v !== 'meridiem');
-    const orderedOtherViews = isRtl ? otherViews.toReversed() : otherViews;
+  const viewsToRender = React.useMemo(() => {
+    if (!isRtl) {
+      return views;
+    }
+    const digitViews = views.filter((v) => v !== 'meridiem');
+    const result = digitViews.toReversed();
+    if (views.includes('meridiem')) {
+      result.push('meridiem');
+    }
+    return result;
+  }, [isRtl, views]);
 
-    // Build other views to be rendered
-    const result = orderedOtherViews.reduce(
+  const viewTimeOptions = React.useMemo(() => {
+    return views.reduce(
       (res, currentView) => {
         return { ...res, [currentView]: buildViewProps(currentView) };
       },
       {} as Record<TimeViewWithMeridiem, MultiSectionDigitalClockViewProps<number>>,
     );
-    // Add built 'meridiem' view at the end, if it was in the original views
-    if (views.includes('meridiem')) {
-      result.meridiem = buildViewProps('meridiem');
-    }
-
-    return result;
-  }, [views, isRtl, buildViewProps]);
+  }, [views, buildViewProps]);
 
   const ownerState = props;
   const classes = useUtilityClasses(ownerState);
@@ -422,11 +423,11 @@ export const MultiSectionDigitalClock = React.forwardRef(function MultiSectionDi
       role="group"
       {...other}
     >
-      {Object.entries(viewTimeOptions).map(([timeView, viewOptions]) => (
+      {viewsToRender.map((timeView) => (
         <MultiSectionDigitalClockSection
           key={timeView}
-          items={viewOptions.items}
-          onChange={viewOptions.onChange}
+          items={viewTimeOptions[timeView].items}
+          onChange={viewTimeOptions[timeView].onChange}
           active={view === timeView}
           autoFocus={autoFocus ?? focusedView === timeView}
           disabled={disabled}
@@ -434,7 +435,7 @@ export const MultiSectionDigitalClock = React.forwardRef(function MultiSectionDi
           slots={slots}
           slotProps={slotProps}
           skipDisabled={skipDisabled}
-          aria-label={localeText.selectViewText(timeView as TimeViewWithMeridiem)}
+          aria-label={localeText.selectViewText(timeView)}
         />
       ))}
```

Comment (arthurbalduini) [reply packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx]: Very elegant ! Done in [0b674d0](https://github.com/mui/mui-x/pull/13447/commits/0b674d0b1b7f7c259f1831aa979cc8d41c5ec0c3)

File: packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx [hunk 4]
```diff
@@ -389,10 +391,22 @@ export const MultiSectionDigitalClock = React.forwardRef(function MultiSectionDi
...
+    }
+    return result;
+  }, [isRtl, views]);
+
   const viewTimeOptions = React.useMemo(() => {
     return views.reduce(
-      (result, currentView) => {
-        return { ...result, [currentView]: buildViewProps(currentView) };
+      (res, currentView) => {
+        return { ...res, [currentView]: buildViewProps(currentView) };
```
Comment (LukasTy) [inline packages/x-date-pickers/src/MultiSectionDigitalClock/MultiSectionDigitalClock.tsx:409]: Nitpick: just to avoid unnecessary diff. 🙈 
```suggestion
      (result, currentView) => {
        return { ...result, [currentView]: buildViewProps(currentView) };
```

**Review by LukasTy — APPROVED** [review]
Great work! 👏 

We are snapshot testing docs only in ltr mode, so, any regressions on this regard might go unnoticed.
WDYT about adding an e2e regression test for `rtl` behavior? 🤔
